linux-gpib-user/doc/man/man3/ibask.3
'\" t
.\"     Title: ibask
.\"    Author: Frank Mori Hess
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 10/04/2025
.\"    Manual: 	Traditional API Functions 
.\"    Source: linux-gpib 4.3.7
.\"  Language: English
.\"
.TH "IBASK" "3" "10/04/2025" "linux-gpib 4.3.7" "Traditional API Functions"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ibask \- query configuration (board or device)
.SH "SYNOPSIS"
.sp
.ft B
.nf
#include <gpib/ib\&.h>
.fi
.ft
.HP \w'int\ ibask('u
.BI "int ibask(int\ " "ud" ", int\ " "option" ", int\ *" "result" ");"
.SH "DESCRIPTION"
.PP
Queries various configuration settings associated with the board or device descriptor
\fIud\fR\&. The
\fIoption\fR
argument specifies the particular setting you wish to query\&. The result of the query is written to the location specified by
\fIresult\fR\&. To change the descriptor\*(Aqs configuration, see
ibconfig()\&.
.PP
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.B Table\ \&5.\ \&ibask options
.TS
allbox expand tab(:);
lB lB lB lB.
T{
option
T}:T{
value (hexadecimal)
T}:T{
result of query
T}:T{
used for board/device
T}
.T&
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l.
T{
IbaPAD
T}:T{
0x1
T}:T{
GPIB primary address
T}:T{
board or device
T}
T{
IbaSAD
T}:T{
0x2
T}:T{
GPIB secondary address (0 for none, 0x60 to 0x7f for secondary
	addresses 0 to 31)
T}:T{
board or device
T}
T{
IbaTMO
T}:T{
0x3
T}:T{
Timeout setting for io operations (a number from 0 to 17)\&.  See
	ibmto()\&.
T}:T{
board or device
T}
T{
IbaEOT
T}:T{
0x4
T}:T{
Nonzero if EOI is asserted with last byte on writes\&.
	See ibeot()\&.
T}:T{
\ \&
T}
T{
IbaPPC
T}:T{
0x5
T}:T{
Parallel poll configuration\&.  See
	ibppc()\&.
T}:T{
board
T}
T{
IbaREADDR
T}:T{
0x6
T}:T{
Useless, included for compatibility only\&.
T}:T{
device
T}
T{
IbaAUTOPOLL
T}:T{
0x7
T}:T{
Nonzero if automatic serial polling is enabled\&.
T}:T{
board
T}
T{
IbaCICPROT
T}:T{
0x8
T}:T{
Useless, included for compatibility only\&.
T}:T{
board
T}
T{
IbaSC
T}:T{
0xa
T}:T{
Nonzero if board is system controller\&.
	See ibrsc()\&.
T}:T{
board
T}
T{
IbaSRE
T}:T{
0xb
T}:T{
Nonzero if board automatically asserts the REN line when
	it becomes the system controller\&.  See
	ibrsc()\&.
T}:T{
board
T}
T{
IbaEOSrd
T}:T{
0xc
T}:T{
Nonzero if termination of reads on reception of the end\-of\-string
	character is enabled\&.  See ibeos(),
	in particular the REOS bit\&.
T}:T{
board or device
T}
T{
IbaEOSwrt
T}:T{
0xd
T}:T{
Nonzero if EOI is asserted whenever end\-of\-string character is
	sent\&.  See ibeos(),
	in particular the XEOS bit\&.
T}:T{
board or device
T}
T{
IbaEOScmp
T}:T{
0xe
T}:T{
Nonzero if all 8 bits are used to match end\-of\-string
	character\&.  Zero if only least significant 7 bits are used\&.
	See ibeos(),
	in particular the BIN bit\&.
T}:T{
board or device
T}
T{
IbaEOSchar
T}:T{
0xf
T}:T{
The end\-of\-string byte\&.
T}:T{
board or device
T}
T{
IbaPP2
T}:T{
0x10
T}:T{
Nonzero if in local parallel poll configure mode\&.  Zero
	if in remote parallel poll configure mode\&.
T}:T{
board
T}
T{
IbaTIMING
T}:T{
0x11
T}:T{
Number indicating T1 delay\&.  1 for 2 microseconds, 2 for 500
	nanoseconds, 3 for 350 nanoseconds\&.  The values are declared in
	the header files as the constants T1_DELAY_2000ns, T1_DELAY_500ns,
	and T1_DELAY_350ns\&.
T}:T{
board
T}
T{
IbaReadAdjust
T}:T{
0x13
T}:T{
Nonzero if byte pairs are automatically swapped during
	reads\&.
T}:T{
board or device
T}
T{
IbaWriteAdjust
T}:T{
0x14
T}:T{
Nonzero if byte pairs are automatically swapped during
	writes\&.
T}:T{
board or device
T}
T{
IbaEventQueue
T}:T{
0x15
T}:T{
Nonzero if event queue is enabled\&.
T}:T{
board
T}
T{
IbaSPollBit
T}:T{
0x16
T}:T{
Nonzero if the use of the SPOLL bit in ibsta is enabled\&.
T}:T{
board
T}
T{
IbaSendLLO
T}:T{
0x17
T}:T{
Nonzero if devices connected to this board are automatically
	put into local lockout
	mode when brought online with ibfind() or ibdev()\&.
T}:T{
board
T}
T{
IbaSPollTime
T}:T{
0x18
T}:T{
Timeout for serial polls\&.  The value of the result
	is between 0 and 17, and has the same meaning as in
	ibtmo()\&.
T}:T{
device
T}
T{
IbaPPollTime
T}:T{
0x19
T}:T{
Timeout for parallel polls\&.  The value of the result
	is between 0 and 17, and has the same meaning as in
	ibtmo()\&.
T}:T{
board
T}
T{
IbaEndBitIsNormal
T}:T{
0x1a
T}:T{
Nonzero if END bit of ibsta is set on reception of end\-of\-string
	character or EOI\&.  Zero if END bit is only set on EOI\&.
T}:T{
board or device
T}
T{
IbaUnAddr
T}:T{
0x1b
T}:T{
Nonzero if UNT (untalk) and UNL (unlisten) commands are automatically
	sent after a completed ibrd* or ibwrt* operation using this descriptor\&.
T}:T{
device
T}
T{
IbaHSCableLength
T}:T{
0x1f
T}:T{
Useless, included only for compatibility\&.
T}:T{
board
T}
T{
IbaIst
T}:T{
0x20
T}:T{
Individual status bit, a\&.k\&.a\&. \*(Aqist\*(Aq\&.
T}:T{
board
T}
T{
IbaRsv
T}:T{
0x21
T}:T{
The current status byte this board will use to respond to
	serial polls\&.
T}:T{
board
T}
T{
IbaBNA
T}:T{
0x200
T}:T{
For a device: the board index (minor number) of interface board through
	which the device is being accessed\&.  For a board: the board index of the
	board itself\&.
T}:T{
board or device
T}
T{
Iba7BitEOS
T}:T{
0x1000
T}:T{
Nonzero if board supports 7 bit EOS comparisons\&.
	See ibeos(),
	in particular the BIN bit\&.  This is a Linux\-GPIB extension\&.
T}:T{
board
T}
.TE
.sp 1
.SH "RETURN VALUE"
.PP
The value of
ibsta
is returned\&.
.SH "COPYRIGHT"
.br
Copyright \(co 2003-2006, 2008 Frank Mori Hess
.br


ChangeLog
2018-12-25 09:50  dpenkler

	* [r1806] Fix board type

2018-12-25 09:46  dpenkler

	* [r1805] Fix ni_usb_gpib cut & paste error

2018-12-25 09:42  dpenkler

	* [r1804] Simplify usb udev rules as per Frank's suggestion

2018-12-21 10:44  dpenkler

	* [r1803] Changes for multiple board configuration

	  Update INSTALL
	  Add serial to device environment
	  Print serial in gpib_udev_config
	  Add example rules for agilent

2018-12-15 08:55  fmhess

	* [r1802] Add some examples for configuring multiple boards with
	  different minor numbers.

2018-12-15 08:29  fmhess

	* [r1801] Automatically gibe the correct board type to gpib_config
	  in udev rules

2018-12-15 08:01  fmhess

	* [r1800] rename gpib generic rules so it runs at 98
	  (before the driver-specific fules which run at 99)

2018-12-15 07:52  fmhess

	* [r1799] Add fix for mylog from gpib_udev_config to
	  gpib_udev_fxloader.in. Also add some quoting to
	  ensure whitespace is preserved in log messages.

2018-12-14 16:23  dpenkler

	* [r1798] Fix logging in gpib_udev_config for gpib_config options
	  --

2018-12-14 08:30  fmhess

	* [r1797] Removed some bash-isms so scripts work with posix
	  /bin/sh.

2018-12-14 08:11  fmhess

	* [r1796] Revert commit 1793

2018-12-14 08:07  fmhess

	* [r1795] Add --with-udev-libdir option to configure, that I failed
	  to include in earlier commit.

2018-12-13 17:05  dpenkler

	* [r1794] Really remove unused files

2018-12-13 17:02  dpenkler

	* [r1793] Add UDEV_LIBDIR to Makefile.ams\nRemove unused files

2018-12-13 13:44  fmhess

	* [r1792] Added --with-udev-libdir configure option, defaulted
	  to PREFIX/lib/udev.

2018-12-13 12:24  dpenkler

	* [r1791] Rename and move udevadm wrapper to libdir

2018-12-13 12:17  fmhess

	* [r1790] Add missing define of UDEV_LIB_DIR to prevent files from
	  getting installed into root directory.

2018-12-13 12:13  fmhess

	* [r1789] Replaced udev scripts ni_usb_gpib and agilent_82357a with
	  a single generic script.

2018-12-13 11:53  fmhess

	* [r1788] Really delete usermap files this time.

2018-12-13 11:47  fmhess

	* [r1787] Pull out firmware loading into a single udev script
	  gpib_udev_fxloader.

2018-12-13 11:44  dpenkler

	* [r1786] Use #!/bin/sh instead of #!/bin/bash

2018-12-13 11:11  dpenkler

	* [r1785] Add wrapper script for udevadm

2018-12-12 18:04  dpenkler

	* [r1784] Fix udev rules scripts for agilent usb

2018-12-12 15:43  fmhess

	* [r1783] Generate usb script/rules during "make all" rather
	  than possibly waiting until "make install", where the
	  user may be running as root.

2018-12-12 15:27  fmhess

	* [r1782] Fixed assignment that was supposed to be a comparison

2018-12-12 14:03  fmhess

	* [r1781] Removed obsolete hotplug usermaps.

	  Fixed problems with usb udev rules not working if the driver
	  module was not already loaded when the hardware was plugged
	  in.

	  Made usb drivers respond to the sysfs device path of their
	  usb interface rather than the parent usb device. This is helpful
	  since the driver binds to the interfaces rather than the device.

2018-12-12 06:25  fmhess

	* [r1780] Update comments in agilent_82357a udev script to use
	  new --serial and --sysfs-device-path gpib_config options.

2018-12-10 10:49  fmhess

	* [r1779] Added linux/atomic.h compatibility header.

2018-12-10 10:36  fmhess

	* [r1778] Added a couple more devices to ni_usb_gpib.usermap.

2018-12-10 10:28  fmhess

	* [r1777] Add kernel driver support for selecting hardware by sysfs
	  device path
	  or serial number.

2018-12-10 10:27  fmhess

	* [r1776] Add support for selecting hardware by sysfs device path
	  or serial number. Updated gpib_config docs with
	  missing info on some options. Noted in docs that
	  gpib.conf will be in the sysconfdir, which isn't
	  always /etc/ depending on configuration.

2018-12-10 10:11  fmhess

	* [r1775] Add device ids for various ni adapters to
	  ni_usb_gpib.usermap

2018-12-07 09:12  dpenkler

	* [r1774] Remove dependency on autoconf tools
	  Modify INSTALL instructions
	  Update drivers/gpib/Makefile
	  Add Makefile to invoke kernel build scripts
	  Add version.h to drivers/gpib/include
	  Make definitions of smp_mb__xxx_atomic independent of kernel
	  version
	  Add VERBOSE flag for make, default is VERBOSE=0
	  so that build log is more overseeable.
	  Add HAVE_DEV_OF_NODE workaround for CentOS/RHEL7
	  Fix compile warnings in lpvo

2018-12-06 06:10  fmhess

	* [r1773] Reverted r1766 since it was actually comparing the serial
	  number to the device tree path. Linux doesn't seem to
	  have a universal (for all bus types) way of specifying
	  a piece of hardware by path, so we can't use the
	  device tree path to specify a usb device anyways.

2018-12-05 11:19  dpenkler

	* [r1772] Apply gpib_lines patch from Tomislav Ivek

2018-11-29 08:29  dpenkler

	* [r1771] Change barrier() to mb() to be generic

2018-11-28 19:02  dpenkler

	* [r1770] Add kernel version ifdef for smp atomics

2018-11-28 18:02  dpenkler

	* [r1769] Add compat fixes for kernels < 3.15 and 3.16

2018-11-28 14:36  dpenkler

	* [r1768] Add patch from Tomislav Ivek

	  The current linux-gpib Python class can run into a double-close
	  situation. This bug is related to a design point in Gpib.Gpib
	  which
	  cleans up its GPIB handle only on Python's GC cycle. This is not
	  guaranteed to run when the programmer expects it to and in
	  extreme
	  cases might even leak resources. This is resolved by adding a
	  Gpib.close() method for user-controlled cleanup.

2018-11-28 10:26  fmhess

	* [r1767] Applied patch from Derek Kozel <derek.kozel@gmail.com>:

	  During the linux-gpib-kernel make install process the kernel
	  modules_install command is run bit the DESTDIR argument is not
	  passed
	  in meaning that it will always try to install to the system
	  directory
	  rather than a custom directory. Adding
	  INSTALL_MOD_PATH="$(DESTDIR)" to
	  the command correctly passes the information.

2018-11-16 20:30  fmhess


linux-gpib-user/lib/ibask.c
/***************************************************************************
                          lib/ibask.c
                             -------------------

	copyright            : (C) 2002 by Frank Mori Hess
    email                : fmhess@users.sourceforge.net
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "ib_internal.h"

int query_ist(const ibBoard_t *board)
{
	int retval;
	struct gpib_board_info_ioctl info;

	retval = ioctl(board->fileno, IBBOARD_INFO, &info);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}

	return info.ist;
}

int query_ppc(const ibBoard_t *board)
{
	int retval;
	struct gpib_board_info_ioctl info;

	retval = ioctl(board->fileno, IBBOARD_INFO, &info);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}

	return info.parallel_poll_configuration;
}

int query_local_ppoll_mode(const ibBoard_t *board)
{
	int retval;
	short local_mode;

	retval = ioctl(board->fileno, IBPP2_GET, &local_mode);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}

	return local_mode;
}

int query_autopoll(const ibBoard_t *board)
{
	int retval;
	struct gpib_board_info_ioctl info;

	retval = ioctl(board->fileno, IBBOARD_INFO, &info);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}

	return info.autopolling;
}

int query_board_t1_delay(const ibBoard_t *board)
{
	int retval;
	struct gpib_board_info_ioctl info;

	retval = ioctl(board->fileno, IBBOARD_INFO, &info);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}

	if (info.t1_delay == 0)	{
		fprintf(stderr, "%s: bug! we don't know what the T1 delay is because it has never been set.\n",
				__FUNCTION__);
		return -EIO;
	} else if (info.t1_delay < 500)
		return T1_DELAY_350ns;
	else if (info.t1_delay < 2000)
		return T1_DELAY_500ns;
	return T1_DELAY_2000ns;
}

int query_board_rsv(const ibBoard_t *board)
{
	int retval;
	int status;

	retval = ioctl(board->fileno, IBQUERY_BOARD_RSV, &status);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}

	return status;
}

int query_pad(const ibBoard_t *board, unsigned int *pad)
{
	int retval;
	struct gpib_board_info_ioctl info;

	retval = ioctl(board->fileno, IBBOARD_INFO, &info);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}

	*pad = info.pad;
	return 0;
}

int query_sad(const ibBoard_t *board, int *sad)
{
	int retval;
	struct gpib_board_info_ioctl info;

	retval = ioctl(board->fileno, IBBOARD_INFO, &info);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}

	*sad = info.sad;
	return 0;
}

int query_no_7_bit_eos(const ibBoard_t *board)
{
	int retval;
	struct gpib_board_info_ioctl info;

	retval = ioctl(board->fileno, IBBOARD_INFO, &info);
	if (retval < 0)	{
		setIberr(EDVR);
		setIbcnt(errno);
		return retval;
	}
	return info.no_7_bit_eos;
}

int ibask(int ud, int option, int *value)
{
	ibConf_t *conf;
	ibBoard_t *board;
	unsigned int pad;
	int sad;
	int retval;

	conf = general_enter_library(ud, 0, 0);
	if (!conf)
		return exit_library(ud, 1);

	board = interfaceBoard(conf);

	switch (option)	{
		case IbaPAD:
			if (conf->is_interface)	{
				retval = query_pad(board, &pad);
				if (retval < 0)
					return exit_library(ud, 1);
				conf->settings.pad = pad;
			}
			*value = conf->settings.pad;
			return exit_library(ud, 0);
			break;
		case IbaSAD:
			if (conf->is_interface)	{
				retval = query_sad(board, &sad);
				if (retval < 0)
					return exit_library(ud, 1);
				conf->settings.sad = sad;
			}
			if (conf->settings.sad < 0)
				*value = 0;
			else
				*value = MSA(conf->settings.sad);
			return exit_library(ud, 0);
			break;
		case IbaTMO:
			*value = usec_to_timeout(conf->settings.usec_timeout);
			return exit_library(ud, 0);
			break;
		case IbaEOT:
			*value = conf->settings.send_eoi;
			return exit_library(ud, 0);
			break;
		case IbaEOSrd:
			*value = conf->settings.eos_flags & REOS;
			return exit_library(ud, 0);
			break;
		case IbaEOSwrt:
			*value = conf->settings.eos_flags & XEOS;
			return exit_library(ud, 0);
			break;
		case IbaEOScmp:
			*value = conf->settings.eos_flags & BIN;
			return exit_library(ud, 0);
			break;
		case IbaEOSchar:
			*value = conf->settings.eos;
			return exit_library(ud, 0);
			break;
		case IbaBNA:
			*value = conf->settings.board;
			return exit_library(ud, 0);
			break;
		case IbaReadAdjust:
			/* XXX I guess I could implement byte swapping stuff,
			 * it's pretty stupid though */
			*value = 0;
			return exit_library(ud, 0);
			break;
		case IbaWriteAdjust:
			/* XXX I guess I could implement byte swapping stuff,
			 * it's pretty stupid though */
			*value = 0;
			return exit_library(ud, 0);
			break;
		case IbaEndBitIsNormal:
			/* XXX no support for setting END status on EOI only yet */
			*value = 1;
			return exit_library(ud, 0);
			break;
		default:
			break;
	}

	if (conf->is_interface)	{
		switch (option) {
			case IbaPPC:
				retval = query_ppc(board);
				if (retval < 0)
					return exit_library(ud, 1);
				*value = retval;
				return exit_library(ud, 0);
				break;
			case IbaAUTOPOLL:
				retval = query_autopoll(board);
				if (retval < 0)
					return exit_library(ud, 1);
				*value = retval;
				return exit_library(ud, 0);
				break;
			case IbaCICPROT:
				// XXX we don't support pass control protocol yet
				*value = 0;
				return exit_library(ud, 0);
				break;
			case IbaIRQ:
				// XXX we don't support interrupt-less operation yet
				*value = 0;
				return exit_library(ud, 0);
				break;
			case IbaSC:
				retval = is_system_controller(board);
				if (retval < 0)
					return exit_library(ud, 1);
				*value = retval;
				return exit_library(ud, 0);
				break;
			case IbaSRE:
				*value = board->set_ren_on_sc;
				return exit_library(ud, 0);
				break;
			case IbaPP2:
				retval = query_local_ppoll_mode(board);
				if (retval < 0)	{
					return exit_library(ud, 1);
				} else {
					*value = retval;
					return exit_library(ud, 0);
				}
				break;
			case IbaTIMING:
				retval = query_board_t1_delay(board);
				if (retval < 0)
					return exit_library(ud, 1);
				*value = retval;
				return exit_library(ud, 0);
				break;
			case IbaDMA:
				// XXX bogus, but pretty unimportant
				*value = board->dma;
				return exit_library(ud, 0);
				break;
			case IbaEventQueue:
				*value = board->use_event_queue;
				return exit_library(ud, 0);
				break;
			case IbaSPollBit:
				*value = 1;
				return exit_library(ud, 0);
				break;
			case IbaSendLLO:
				*value = conf->settings.local_lockout;
				return exit_library(ud, 0);
				break;
			case IbaPPollTime:
				*value = usec_to_ppoll_timeout(conf->settings.ppoll_usec_timeout);
				return exit_library(ud, 0);
				break;
			case IbaHSCableLength:
				/* HS transfer not supported and may never
				 * be as it is not part of GPIB standard */
				*value = 0;
				return exit_library(ud, 0);
				break;
			case IbaIst:
				retval = query_ist(board);
				if (retval < 0)
					return exit_library(ud, 1);
				*value = retval;
				return exit_library(ud, 0);
				break;
			case IbaRsv:
				retval = query_board_rsv(board);
				if (retval < 0)
					return exit_library(ud, 1);
				*value = retval;
				return exit_library(ud, 0);
				break;
			case Iba7BitEOS:
				retval = query_no_7_bit_eos(board);
				if (retval < 0)
					return exit_library(ud, 1);
				*value = !retval;
				return exit_library(ud, 0);
				break;
			default:
				break;
		}
	} else {
		switch (option)	{
			case IbaREADDR:
				*value = conf->settings.readdr;
				return exit_library(ud, 0);
				break;
			case IbaSPollTime:
				*value = usec_to_timeout(conf->settings.spoll_usec_timeout);
				return exit_library(ud, 0);
				break;
			case IbaUnAddr:
				*value = conf->settings.send_unt_unl;
				return exit_library(ud, 0);
				break;
			default:
				break;
		}
	}

	setIberr(EARG);

	return exit_library(ud, 1);
}


